File: type-parameters/rotate_array.py
from typing import TypeVar


T = TypeVar('T')


def rotate_array_left(array: list[T], rotation: int) -> list[T]:
    result: list[T] = []
    for i in range(len(array)):
        new_index = (i + rotation) % len(array)
        result.append(array[new_index])
    return result


strings = ['is', 'cool', 'hello', 'world!', 'python']
strings_rotated = rotate_array_left(strings, 2)
for s in strings_rotated:
    print(s)

File: 03-type-parameters/rotate_array.cpp
#include <cstddef>
#include <iostream>
#include <string>
#include <vector>


template<typename T>
std::vector<T> rotate_array_left(std::vector<T> const& array, unsigned rotation) {
    std::vector<T> result;
    for (std::size_t i = 0; i < array.size(); ++i) {
        auto const new_index = (i + rotation) % array.size();
        result.push_back(array[new_index]);
    }
    return result;
}


int main() {
    std::vector<std::string> const strings{"is", "cool", "hello", "world!", "C++"};
    auto const strings_rotated = rotate_array_left(strings, 2);
    for (auto const& s : strings_rotated) {
        std::cout << s << std::endl;
    }
}
